libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -268,8 +268,137 @@ string IRGenerator::generateFunction(FunctionDefinition const& _function)
 			retParams += m_context.addLocalVariable(*varDecl).stackSlots();
 			retInit += generateInitialAssignment(*varDecl);
 		}
+
+		t("retParams", joinHumanReadable(retParams));
+		t("retInit", retInit);
+
+		if (_function.modifiers().empty())
+			t("body", generate(_function.body()));
+		else
+		{
+			for (size_t i = 0; i < _function.modifiers().size(); ++i)
+			{
+				ModifierInvocation const& modifier = *_function.modifiers().at(i);
+				string next =
+					i + 1 < _function.modifiers().size() ?
+					IRNames::modifierInvocation(*_function.modifiers().at(i + 1)) :
+					IRNames::functionWithModifierInner(_function);
+				generateModifier(modifier, _function, next);
+			}
+			t("body",
+				(retParams.empty() ? string{} : joinHumanReadable(retParams) + " := ") +
+				IRNames::modifierInvocation(*_function.modifiers().at(0)) +
+				"(" +
+				joinHumanReadable(retParams + params) +
+				")"
+			);
+			// Now generate the actual inner function.
+			generateFunctionWithModifierInner(_function);
+		}
+		return t.render();
+	});
+}
+
+string IRGenerator::generateModifier(
+	ModifierInvocation const& _modifierInvocation,
+	FunctionDefinition const& _function,
+	string const& _nextFunction
+)
+{
+	string functionName = IRNames::modifierInvocation(_modifierInvocation);
+	return m_context.functionCollector().createFunction(functionName, [&]() {
+		m_context.resetLocalVariables();
+		Whiskers t(R"(
+			function <functionName>(<params>)<?+retParams> -> <retParams></+retParams> {
+				<assignRetParams>
+				<evalArgs>
+				<body>
+			}
+		)");
+		t("functionName", functionName);
+		vector<string> retParamsIn;
+		for (auto const& varDecl: _function.returnParameters())
+			retParamsIn += IRVariable(*varDecl).stackSlots();
+		vector<string> params = retParamsIn;
+		for (auto const& varDecl: _function.parameters())
+			params += m_context.addLocalVariable(*varDecl).stackSlots();
+		t("params", joinHumanReadable(params));
+		vector<string> retParams;
+		string assignRetParams;
+		for (size_t i = 0; i < retParamsIn.size(); ++i)
+		{
+			retParams.emplace_back(m_context.newYulVariable());
+			assignRetParams += retParams.back() + " := " + retParamsIn[i] + "\n";
+		}
 		t("retParams", joinHumanReadable(retParams));
-		t("initReturnVariables", retInit);
+		t("assignRetParams", assignRetParams);
+
+		solAssert(*_modifierInvocation.name().annotation().requiredLookup == VirtualLookup::Virtual, "");
+
+		ModifierDefinition const& modifier = dynamic_cast<ModifierDefinition const&>(
+			*_modifierInvocation.name().annotation().referencedDeclaration
+		).resolveVirtual(m_context.mostDerivedContract());
+
+		solAssert(
+			modifier.parameters().empty() ==
+			(!_modifierInvocation.arguments() || _modifierInvocation.arguments()->empty()),
+			""
+		);
```

Comment:
Maybe better to move at the beginning of a function?

Reply:
I put it here so that it is right before where the arguments are processed.